### ISSUE TITLE
research: train and validate a hybrid physics-residual pedestrian goal estimator

### ISSUE THREAD

**Issue by ll7** [ll7/robot_sf_ll7#8081]
## Current issue-planning review — 2026-09-15

**Audit disposition:** `blocked`. The current audit classifies the issue as blocked; its prerequisite, evidence, or triage gate is incomplete.

**First action:** Train and select a small hybrid learned estimator that corrects and calibrates the observation-only physics/rule baseline rather than relearning the entire pedestrian dynamics model from raw trajectories.

### Current bounded plan

1. Inspect the named owners and dependencies before editing; keep the live issue scope and fail-closed evidence boundary intact.
2. Run the named focused validation, record exact output and remaining blockers, and keep fallback, degraded, unavailable, and inconclusive outcomes explicit.

### Files and owners to inspect

- docs/maintainer_values.md
- docs/context/issue_1512_issue_archetypes.md
- docs/context/pedestrian_goal_force_prediction_research.md
- tests/training/test_goal_belief_models.py
- tests/training/test_goal_belief_losses.py
- tests/training/test_goal_belief_masking.py
- tests/training/test_goal_belief_training_determinism.py
- tests/prediction/test_learned_goal_belief_inference.py
- tests/prediction/test_learned_goal_belief_calibration.py
- tests/models/test_goal_belief_checkpoint_registry.py
- tests/benchmark/test_goal_belief_model_comparison.py
- configs/training/goal_belief_model_selection.yaml
- configs/benchmarks/goal_belief_model_selection_eval.yaml
- model/artifact

### Acceptance and validation

- Demonstrate the stated outcome from the live issue with a focused, reproducible check.
- Keep any missing prerequisite or inconclusive result explicit rather than treating it as success.

### Current dependencies and issue-text corrections

- No supplied individual draft existed for this issue in the handoff package; this is a conservative plan derived from the refreshed live body, and the historical issue text is preserved below.
- Live issue references to verify: https://github.com/ll7/robot_sf_ll7/issues/8060; https://github.com/ll7/robot_sf_ll7/issues/8077; https://github.com/ll7/robot_sf_ll7/issues/8078; https://github.com/ll7/robot_sf_ll7/issues/8079; https://github.com/ll7/robot_sf_ll7/issues/8068; https://github.com/ll7/robot_sf_ll7/issues/8072; https://github.com/ll7/robot_sf_ll7/issues/8075; https://github.com/ll7/robot_sf_ll7/issues/8076; https://github.com/ll7/robot_sf_ll7/issues/8058
- Keep current out-of-scope, ownership, provenance, rights, compute, and claim boundaries intact. Do not close the issue merely because this plan was posted.

### Lifecycle and claim boundary

- Current state labels: `state:blocked`.
- The issue remains open. No closure, compute submission, release/publication, evidence deletion, or scientific-claim promotion is authorized by this review.
- This review does not request a lifecycle transition; existing assignees, ownership, valid priority labels, and unrelated labels are preserved.
- Proposed non-lifecycle label additions applied with this review: `none`.

**Evidence note:** Full live issue body and complete available comment thread were reviewed during the 2026-09-15 refresh. This is planning metadata, not implementation or benchmark evidence; historical measurements remain historical.

## Archetype Metadata

```yaml
archetype: training-campaign
evidence_tier: nominal
linked_policy:
  - docs/maintainer_values.md
  - docs/context/issue_1512_issue_archetypes.md
  - docs/context/pedestrian_goal_force_prediction_research.md
```

## Parent and dependencies

- Parent programme: #8060
- Hard predecessors: #8077, #8078, #8079
- Requires completed rule baselines: #8068, #8072, #8075, #8076
- Research/design basis: PR #8058
- Unblocks: bounded online adaptation, compact observation admission, and paired closed-loop preregistration

## Goal

Train and select a small hybrid learned estimator that corrects and calibrates the observation-only physics/rule baseline rather than relearning the entire pedestrian dynamics model from raw trajectories.

The selected model should predict, for each tracked pedestrian:

1. residual correction to goal-force mean;
2. force uncertainty or a bounded multimodal distribution;
3. active-waypoint and final-route/destination probabilities, including unknown;
4. arrival/stopped/restart probabilities;
5. goal-change probability and event subtype support;
6. optional preferred-speed/relaxation-time posterior;
7. optional forecast residual or next-state auxiliary output.

The campaign must compare simple models before recurrent/graph models, use grouped held-out partitions, preserve every negative arm, and apply Gate E using proper scores, calibration, change quality, and runtime—not training loss alone.

## Research questions

1. Does learned residual correction improve observation-only goal-force NLL and direction error over the best rule baseline?
2. Does temporal recurrence improve change detection and route inference beyond explicit multi-window Bayes?
3. Does graph neighbour context help primarily under interactions or merely overfit simulator layouts?
4. Which history lengths carry useful information after accounting for runtime and reset lag?
5. Can a learned model retain calibrated uncertainty under unseen maps/routes, hidden force contributors, parameter shifts, and absent candidates?
6. Do open-loop forecast gains from #8078 increase with the learned belief?

## Competing explanations

- The rule baseline already captures all useful structure; learned residuals do not transfer.
- Improvements come from oracle leakage or split contamination.
- A non-temporal model is sufficient; recurrence adds complexity without value.
- Graph context helps only in in-domain dense scenes.
- Point error improves while probabilistic calibration worsens.
- Learned change probability merely follows raw heading turns and increases false resets.
- Gains disappear when tracking/candidate errors are included rather than oracle-clean subsets.

## Frozen inputs

Before execution, bind:

- dataset manifest and per-shard checksums from #8079;
- train/validation/held-out/stress split digests;
- feature and label schema versions;
- rule-baseline version/config/digest;
- candidate and tracking configurations;
- normalizer fit receipt;
- model family grid;
- seeds and deterministic environment settings;
- training budget per arm;
- primary/secondary metrics and selection rule;
- fallback/exclusion policy;
- output/model registry destination.

No manifest or split mutation after the first model sees held-out confirmation data. Any necessary correction creates a successor dataset version and restarts selection.

## Model ladder

Train and evaluate in this order:

### M0 — non-learned references

- zero/no residual correction;
- best rule-based `GoalBelief` selected by #8077;
- oracle-component upper bound, evaluator only.

### M1 — ridge/linear residual model

- current-window fixed features;
- goal-force residual and optional candidate logits;
- strong interpretability/control baseline.

### M2 — gradient-boosted or tree baseline

Only if an existing lightweight dependency is already supported or can be added with clear value. Otherwise document as unavailable; do not add a heavy dependency solely for this arm.

### M3 — small MLP

- fixed multi-window summary features;
- multi-head output;
- tests whether nonlinear current-state correction suffices.

### M4 — small GRU

- variable causal history with explicit masks;
- shared recurrent state plus typed output heads;
- reset-aware sequence boundaries.

### M5 — graph-GRU

- target pedestrian, visible neighbours, and robot context;
- deterministic node ordering/masks;
- message passing or pooled node context followed by GRU;
- no oracle group/identity information.

### M6 — larger graph/Transformer model

Only admissible if M5 leaves a demonstrated, statistically and practically material held-out gap and runtime remains feasible. This is not a default deliverable.

## Recommended first production candidate

A small physics-residual GRU should be the expected candidate, not a mandated winner. Its input is the actor-visible feature sequence from #8079. Its output augments the rule belief:

```text
f_goal_learned = f_goal_rule + delta_f_network
```

Uncertainty head options:

- Cholesky-parameterized 2×2 Gaussian covariance;
- small K-mode Gaussian mixture when validation proves multimodality is needed;
- circular direction distribution plus magnitude distribution.

Covariance must remain positive definite through a stable parameterization. Clamp ranges and numerical floors must be explicit and tested.

## Input ablations

At minimum compare:

1. kinematics only;
2. kinematics + rule goal-force;
3. + typed reconstructed force components;
4. + candidate/path-tangent features;
5. + hierarchical rule posterior;
6. + change/reset diagnostics;
7. + neighbour graph context;
8. known dynamics parameters versus actor priors;
9. `H={1,2,3,5,8,16}` and random-history model;
10. no-unknown versus explicit unknown is prohibited as a production arm but may be a diagnostic ablation.

No ablation may use oracle goal/route/force components as actor inputs. Oracle feature arms, if included for upper-bound research, must have distinct model IDs and never be planner-admissible.

## Output heads and losses

Use a named, mask-aware multi-task objective. Candidate form:

```text
L = lambda_force * NLL(true goal force | predicted distribution)
  + lambda_direction * circular/cosine loss
  + lambda_waypoint * CE(active waypoint candidate | representable rows)
  + lambda_route * CE(final route/destination | representable rows)
  + lambda_unknown * BCE(candidate missing)
  + lambda_change * focal/BCE/CE(change event)
  + lambda_arrival * CE(arrival/stop/restart)
  + lambda_dynamics * Huber(next velocity or forecast residual)
  + lambda_calibration * declared calibration regularizer.
```

Requirements:

- every head has an explicit valid-row mask and denominator;
- censored force magnitude is excluded or uses a valid censored likelihood;
- candidate-missing rows supervise unknown, not a wrong nearest candidate;
- class weighting is derived from training prevalence only;
- loss weights are frozen before held-out confirmation;
- report raw per-head losses and metrics, not only weighted total;
- verify that auxiliary heads do not degrade the primary proper score through an ablation.

## Training protocol

- multiple fixed seeds per serious model family;
- early stopping on validation primary score with declared patience/minimum delta;
- identical maximum examples/epochs or a declared matched-compute budget;
- no test/held-out feedback during hyperparameter search;
- deterministic CPU smoke before GPU/SLURM launch;
- optimizer, scheduler, precision, clipping, batch size, and thread environment recorded;
- checkpoint every selected/best/final state with digest;
- retain failed/diverged runs in the run ledger;
- no silent automatic fallback to an untrained/constant-velocity model.

## Calibration

Evaluate raw and post-hoc calibrated variants. Allowed post-hoc methods may include temperature scaling for categorical heads and covariance scaling/isotonic-like calibration for continuous uncertainty, fitted on validation only.

Report:

- calibration before and after adjustment;
- held-out NLL/Brier/energy score;
- coverage at multiple nominal levels;
- sharpness;
- calibration by interaction/event/candidate/tracking strata;
- calibration transfer to stress partitions;
- calibrator digest and feature/model compatibility.

Do not choose a model that improves point error by becoming unjustifiably sharp.

## Held-out and stress evaluation

Use #8077 on:

- held-out maps;
- held-out scenario families;
- unseen route branches/destinations;
- goal missing from candidate set;
- waypoint progression and redirect patterns;
- stopping/restart;
- tracking reorder/occlusion/reacquisition;
- hidden force contributors;
- preferred-speed/relaxation-time perturbations;
- alternate supported pedestrian models;
- sparse and dense crowds;
- speed saturation;
- rotated/translated geometric copies.

Report all failures and availability denominators. An oracle-clean subset may diagnose upper bounds but cannot replace actor-condition results.

## Change/reset evaluation

A learned change head must be evaluated both:

1. as a score without altering the rule controller; and
2. in a shadow learned-controller configuration.

Only if held-out false-reset/delay metrics meet the predeclared Gate D/E limits may it become the default detector. Otherwise retain the rule change controller and use the learned score as an auxiliary diagnostic.

## Open-loop forecast propagation

Feed every selected candidate model through #8078 without changing planner actions. Report whether improvements in force/route scores propagate to forecast NLL/ADE/FDE/coverage. A model with better goal labels but worse trajectory forecasts requires explicit analysis and cannot advance automatically.

## Selection rule — Gate E

Predeclare a primary comparator and selection rule. A model may advance only if all conditions hold:

1. actor leakage/split audits pass;
2. held-out goal-force or direction proper score improves over the best rule baseline by a predeclared meaningful margin, or achieves non-inferiority with materially better change/route behavior;
3. route/waypoint/unknown scores do not materially regress;
4. false reset/detection delay remains within Gate D bounds;
5. calibration coverage remains within declared tolerance;
6. open-loop forecast proper score is non-inferior or better;
7. p95 runtime and memory fit the declared planner budget;
8. no fallback/degraded/untrained mode contributes to nominal rows;
9. gains are not confined to one map or oracle-clean subset;
10. exact checkpoint/normalizer/config is reproducible and registry-valid.

If no learned model passes, select the best rule baseline. That is a valid negative result.

## Required tests

Suggested files:

- `tests/training/test_goal_belief_models.py`
- `tests/training/test_goal_belief_losses.py`
- `tests/training/test_goal_belief_masking.py`
- `tests/training/test_goal_belief_training_determinism.py`
- `tests/prediction/test_learned_goal_belief_inference.py`
- `tests/prediction/test_learned_goal_belief_calibration.py`
- `tests/models/test_goal_belief_checkpoint_registry.py`
- `tests/benchmark/test_goal_belief_model_comparison.py`

Required cases:

- output shapes and positive-definite covariance;
- mixture weights normalized;
- masks exclude censored/unrepresented rows from incompatible losses;
- reset boundaries clear recurrent state;
- variable-history masks and `H=1` no-history behavior;
- deterministic tiny training and checkpoint round-trip;
- oracle/future feature canaries;
- feature schema/checkpoint/normalizer mismatch fails closed;
- missing checkpoint does not silently become nominal fallback;
- calibration fitted only on validation;
- report includes all seeds and failed runs;
- selected model exactly matches selection rule;
- rule baseline remains available and unchanged.

## Execution commands

The issue must freeze canonical commands through current training infrastructure, conceptually:

```bash
uv run python -m robot_sf.training.goal_belief_train \
  --config configs/training/goal_belief_model_selection.yaml \
  --manifest <frozen-dataset-manifest> \
  --output output/models/goal_belief_model_selection

uv run python -m robot_sf.benchmark.goal_belief_eval run \
  --config configs/benchmarks/goal_belief_model_selection_eval.yaml \
  --models output/models/goal_belief_model_selection/model_index.json \
  --output output/benchmarks/goal_belief_model_selection

uv run python -m robot_sf.training.goal_belief_selection validate \
  --report output/benchmarks/goal_belief_model_selection/report.json
```

GPU/SLURM runs require a launch packet and exact resource budget. Local smoke must pass first.

## Evidence outputs

Produce durable, versioned:

- frozen campaign config and dataset manifest references;
- run ledger with all arms/seeds/statuses;
- checkpoint manifests and SHA-256 digests;
- normalizer/calibrator digests;
- train/validation curves;
- held-out/stress evaluation report;
- open-loop forecast propagation report;
- runtime/memory report;
- selection decision and rejected-arm reasons;
- model card with actor inputs, limitations, fallback behavior, and claim boundary;
- independent report-validation receipt.

Commit compact reports/manifests only. Store large checkpoints through the existing model/artifact policy and register exact paths/digests.

## Acceptance criteria

- [ ] M0–M4 are evaluated; M5 is evaluated if graph features are complete; M6 requires an explicit gap justification.
- [ ] Every model uses actor-only causal features and grouped leakage-free splits.
- [ ] Loss masks and denominators are correct for censored, unknown, and unavailable labels.
- [ ] Multiple seeds, failed runs, and matched budgets are recorded.
- [ ] Calibration and proper scores accompany point errors.
- [ ] Held-out maps/scenarios and stress conditions are evaluated.
- [ ] Change/reset and open-loop forecast propagation are checked.
- [ ] Runtime and memory are measured at intended scene scale.
- [ ] Checkpoint, normalizer, calibrator, config, and schema are hash-bound.
- [ ] Selection follows the preregistered Gate E rule.
- [ ] If no learned model passes, the issue records a negative result and keeps the rule baseline.

## Falsification and stop rule

Stop learned-model promotion and record negative/inconclusive if:

- gains vanish on held-out or actor-condition subsets;
- proper scores/calibration worsen despite lower point error;
- split/leakage audit fails;
- learned change head exceeds false-reset bounds;
- forecast quality regresses;
- runtime exceeds budget;
- selected checkpoint cannot be deterministically reconstructed/validated;
- only oracle-feature or fallback arms improve.

Do not modify the held-out set, thresholds, or primary metric after seeing the result without starting a new versioned campaign.

## Risks and rollback

- Simulator overfit: grouped transfer partitions and alternate-model stress.
- Multi-task negative transfer: per-head ablations and a force-only residual baseline.
- Class imbalance: train-derived weighting and natural-distribution evaluation.
- Uncertainty collapse: proper scoring and post-hoc validation-only calibration.
- Recurrent state leakage: explicit segment reset tests.
- Rollback: unregister/reject learned checkpoint and retain the best rule-based estimator.

## Downstream propagation

Update #8060 with Gate E result, exact selected estimator/checkpoint or the negative rule-baseline decision. No planner ranking, release, paper, or dissertation claim follows until a separate paired closed-loop preregistration and execution pass Gate F.

```yaml
schema: goal_autopilot_preparation.v1
repository: ll7/robot_sf_ll7
issue: 8081
source_body_sha256: 59007ad458720caea593e24837e21878ffe56127160e0625475dd1f40fdc5e35
source_comments_sha256: 
audit_schema: open_issue_contract_audit.v1
audit_digest: 1c191411793929640684e0827d763f3a2009a8682ab98dcefa8c2ded3e54f4ea
audit_classification: blocked
next_action: resolve_named_blocker
authority: blocker_owner
execution_mode: blocker
preferred_worker: MaxRunner
expected_pr_runner_label: runner:max
implementation_admitted: False
state_ready_change_proposed: False
mutation_batch: open-issues-20260830-06
```

This packet is preparation evidence only. It never overrides live labels, exact claim state, branch state, typed dependencies, domain gates, compute authority, release authority, or scientific evidence rules.
